src/System.Net.HttpListener/tests/HttpListenerResponseTests.cs

```diff
@@ -53,13 +53,1208 @@ public void Dispose()
+                }
+                catch (HttpListenerException)
+                {
+                    Assert.False(willBlock);
+                }
+            }
+        }
+
+        [ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotOneCoreUAP))]
+        public async Task Headers_GetSet_ReturnsExpected()
+        {
+            using (HttpListenerResponse response = await GetResponse())
+            {
+                Assert.Empty(response.Headers);
+                response.Headers.Add("OldName", "OldValue");
+
+                var headers = new WebHeaderCollection
+                {
+                    { "Name1", "Value1" },
+                    { "Name2", "Value2" },
+                    { "Name3", "" }
+                };
+                response.Headers = headers;
+                Assert.Equal(headers, response.Headers);
+            }
+
+            string clientResponse = GetClientResponse();
+            Assert.Contains("\r\nName1: Value1\r\nName2: Value2\r\nName3: \r\n", clientResponse);
+        }
+
+        [ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotOneCoreUAP))]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, ".NET Core fixes a bug where HttpListenerResponse.CopyFrom(null) throws a NRE.")]
+        // The managed implementation should store ContentType in Headers, not as a separate variable.
+        [ActiveIssue(18128, TestPlatforms.AnyUnix)]
+        public async Task Headers_SetNull_ThrowsArgumentNullExceptionInNetCore()
+        {
+            using (HttpListenerResponse response = await GetResponse())
+            {
+                response.Headers.Add("OldName", "OldValue");
+
+                Assert.Throws<ArgumentNullException>("value", () => response.Headers = null);
+                Assert.Equal(1, response.Headers.Count);
+            }
+        }
+
+        [ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotOneCoreUAP))]
+        [SkipOnTargetFramework(~TargetFrameworkMonikers.NetFramework, ".NET Core fixes a bug where HttpListenerResponse.CopyFrom(null) throws a NRE.")]
+        public async Task Headers_SetNull_ThrowsNullReferenceExceptionInNetFx()
+        {
+            using (HttpListenerResponse response = await GetResponse())
+            {
+                response.Headers.Add("OldName", "OldValue");
+
+                Assert.Throws<NullReferenceException>(() => response.Headers = null);
+                Assert.Equal(0, response.Headers.Count);
+            }
+        }
+
+        private string GetClientResponse()
+        {
+            byte[] buffer = new byte[256];
+            int bytesReceived = Client.Receive(buffer);
+            Assert.True(bytesReceived < buffer.Length, "Buffer should hold enough space to hold the response from the server.");
+
+            return Encoding.UTF8.GetString(buffer, 0, bytesReceived);
```

Comment:
The reason why I went with Socket instead of HttpClient is because HttpClient doesn't seem to read some of the headers that are sent back from the HttpListener - this could be because they are invalid, for example. Socket allows me to actually verify we send what we expect. HttpClient simply sets headers it doesn't like to null, for example.

I thought that doing synchronous writes and reads would avoid this issue - I can look into using HttpClient again, but it'd be nice to have some kind of way to read the exact response from the server.